### PR TITLE
add rubocop

### PR DESCRIPTION
**Pull request by evpgh**
https://github.com/ronin-rb/ronin-web/issues/33
- Add rubocop
- Add rubocop to CI

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+# Style/StringLiterals: { EnforcedStyle: double_quotes }
+# Style/WordArray
+
+# rubocop cannot tell that rubygems_mfa_required is enabled in gemspec.yml
+Gemspec/RequireMFA: { Enabled: false }
+
+# make an exception for our gemspec code
+Gemspec/DuplicatedAssignment:
+  Enabled:
+    - 'ronin-web-template.gemspec'
```
Comment (postmodern) [inline .rubocop.yml:80]: `ronin-web-template` -> `ronin-web-browser`.

File: Gemfile [hunk 2]
```diff
@@ -18,10 +19,12 @@ group :development do
   gem 'kramdown',      '~> 2.0'
   gem 'kramdown-man',  '~> 0.1'
 
-  gem 'redcarpet',       platform: :mri
-  gem 'yard',           '~> 0.9'
+  gem 'redcarpet', platform: :mri
```
Comment (postmodern) [inline Gemfile:22]: I'm actually a fan of vertical alignment. It makes it easier to edit the version requirements.

File: Rakefile [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
-  warn "Run `bundle install` to install missing gems"
+  warn 'Run `bundle install` to install missing gems'
   exit e.status_code
 end
 
 require 'rake'
 
 require 'rubygems/tasks'
-Gem::Tasks.new(sign: {checksum: true, pgp: true})
+Gem::Tasks.new(sign: { checksum: true, pgp: true })
```
Comment (postmodern) [inline Rakefile:23]: Is `rubocop` adding those spaces?

File: lib/ronin/web/browser/agent.rb [hunk 4]
```diff
@@ -58,22 +58,22 @@ def initialize(visible:  false,
                   when Hash, nil then proxy
                   when URI::HTTP, Addressable::URI
                     {
-                      host:     proxy.host,
-                      port:     proxy.port,
-                      user:     proxy.user,
+                      host: proxy.host,
+                      port: proxy.port,
+                      user: proxy.user,
                       password: proxy.password
```
Comment (postmodern) [inline lib/ronin/web/browser/agent.rb:64]: Again, I prefer vertical alignment.

File: lib/ronin/web/browser/agent.rb [hunk 5]
```diff
@@ -58,22 +58,22 @@ def initialize(visible:  false,
...
-                      port:     uri.port,
-                      user:     uri.user,
+                      host: uri.host,
+                      port: uri.port,
+                      user: uri.user,
                       password: uri.password
                     }
                   else
-                    raise(ArgumentError,"invalid proxy value (#{proxy.inspect}), must be either a Hash, URI::HTTP, String, or nil")
+                    raise(ArgumentError, "invalid proxy value (#{proxy.inspect}), must be either a Hash, URI::HTTP, String, or nil")
```
Comment (postmodern) [inline lib/ronin/web/browser/agent.rb:76]: I tend to omit the spaces in method calls. It saves time and doesn't impact readability that much (I think).

File: lib/ronin/web/browser.rb [hunk 6]
```diff
@@ -139,8 +140,8 @@ def self.new(**kwargs)
       # @return [Agent]
       #   A new instance of a headless or visible Chrome browser.
       #
-      def self.open(**kwargs,&block)
-        Agent.open(**kwargs,&block)
+      def self.open(**kwargs, &block)
+        Agent.open(**kwargs, &block)
```
Comment (postmodern) [inline lib/ronin/web/browser.rb:144]: It's OK to omit spaces between arguments here.

File: lib/ronin/web/browser/agent.rb [hunk 7]
```diff
@@ -168,18 +168,16 @@ def proxy?
         #
         # @yieldparam [Integer] total
         #
-        def on(event,&block)
+        def on(event, &block)
           case event
           when :response
-            super('Network.responseReceived') do |params,index,total|
+            super('Network.responseReceived') do |params, index, total|
```
Comment (postmodern) [inline lib/ronin/web/browser/agent.rb:174]: Not a fan of spaces in block arguments.

File: ronin-web-browser.gemspec [hunk 8]
```diff
@@ -36,26 +36,22 @@ Gem::Specification.new do |gem|
...
+  gemspec['dependencies']&.each do |name, versions|
+    gem.add_dependency(name, split[versions])
   end
 
-  if gemspec['development_dependencies']
-    gemspec['development_dependencies'].each do |name,versions|
-      gem.add_development_dependency(name,split[versions])
-    end
+  gemspec['development_dependencies']&.each do |name, versions|
+    gem.add_development_dependency(name, split[versions])
```
Comment (postmodern) [inline ronin-web-browser.gemspec:55]: I would just ignore the `.gemspec` file. It's a boilerplate file that can be updated later by coping in a new version.

**Comment by postmodern** [comment]
I think you forgot to copy in the full `.rubocop.yml` file from `ronin-template`. A lot of the rules in it disable [various pedantic rubocop rules](https://github.com/ronin-rb/ronin-template/blob/main/.rubocop.yml#L88-L155). I can also review the disabled rules and ensure that they should be left disabled.

File: spec/agent_spec.rb [hunk 9]
```diff
@@ -1,16 +1,17 @@
+# frozen_string_literal: true
 require 'spec_helper'
 require 'ronin/web/browser/agent'
 
 describe Ronin::Web::Browser::Agent do
-  describe "#initialize" do
-    context "when Ronin::Support::Network::HTTP.proxy is set" do
+  describe '#initialize' do
+    context 'when Ronin::Support::Network::HTTP.proxy is set' do
```
Comment (postmodern) [inline spec/agent_spec.rb:7]: For some reason I always use double quoted strings in RSpec tests. Old habits die hard I guess.

File: spec/browser_spec.rb [hunk 10]
```diff
@@ -79,16 +80,16 @@
...
-  describe ".open" do
-    context "when given a block" do
+  describe '.open' do
+    context 'when given a block' do
       it "must yield a new #{described_class} instance" do
-        expect { |b|
+        expect do |b|
           subject.open(&b)
-        }.to yield_successive_args(described_class::Agent)
+        end.to yield_successive_args(described_class::Agent)
```
Comment (postmodern) [inline spec/browser_spec.rb:88]: I **strongly** disagree with this rubocop rule. When chaining together multiple methods that take a block, curly brace blocks are the defacto standard. I really have no clue how that rule got into rubocop in the first place, and wish they would remove it.

File: .rubocop.yml [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+# Style/KeywordParametersOrder
+# Style/MethodCallWithoutArgsParentheses
+# Style/MutableConstant
+# Style/QuotedSymbols: { EnforcedStyle: double_quotes }
+# Style/RedundantReturn
+# Style/SafeNavigation
+# Style/SpecialGlobalVars
+# Style/StringLiterals: { EnforcedStyle: double_quotes }
+# Style/WordArray
+
```
Comment (postmodern) [inline .rubocop.yml:73]: Missing the other disabled rules from the [`.rubocop.yml` template](https://github.com/ronin-rb/ronin-template/blob/b107a4b0e0e6e7836c88f4cd26d1522e90c8d989/.rubocop.yml#L88-L155).

File: lib/ronin/web/browser/agent.rb [hunk 12]
```diff
@@ -106,7 +106,7 @@ def self.open(**kwargs)
             browser.quit
           end
 
-          return browser
+          browser
```
Comment (postmodern) [inline lib/ronin/web/browser/agent.rb:109]: I'm a fan of sometimes using explicit `return`s. I feel it's weird to have a local variable just sitting there on the last line. I know, it's a weird habit that I developed after years of writing C/C++/Java...

**Review by postmodern — CHANGES_REQUESTED** [review]
Missing the other disabled rubocop rules from the [`.rubocop.yml` template](https://github.com/ronin-rb/ronin-template/blob/b107a4b0e0e6e7836c88f4cd26d1522e90c8d989/.rubocop.yml#L88-L155), which should eliminate most of the small changes.

**Comment by evpgh** [comment]
I'll just close this PR and create a new one.

**Comment by postmodern** [comment]
Let me create the `rubocop-ronin` gem to contain the new common rubocop rules and update the template `.rubocop.yml` to use that.

**Comment by postmodern** [comment]
Done. See https://github.com/ronin-rb/rubocop-ronin#readme and the new [`.rubocop.yml` template file](https://github.com/ronin-rb/ronin-template/blob/main/.rubocop.yml).

**Pull request closed without merging**